src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -690,11 +690,18 @@ protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement n
             if (node.ExpressionOpt != null && this.State.Reachable)
             {
                 TypeSymbolWithAnnotations returnType = this._currentMethodOrLambda?.ReturnType;
-
-                if (_result.Type?.IsNullable == true)
+                bool returnTypeIsNonNullable = IsNonNullable(returnType);
+                bool returnTypeIsUnconstrainedTypeParameter = IsUnconstrainedTypeParameter(returnType?.TypeSymbol);
+                bool reportedNullable = false;
+                if (returnTypeIsNonNullable || returnTypeIsUnconstrainedTypeParameter)
+                {
+                    reportedNullable = CheckNullAsNonNullableReference(node.ExpressionOpt);
+                }
+                if (!reportedNullable)
                 {
-                    if ((object)returnType != null && returnType.IsReferenceType && returnType.IsNullable == false &&
-                        !CheckNullAsNonNullableReference(node.ExpressionOpt))
+                    TypeSymbolWithAnnotations resultType = _result.Type;
+                    if (IsNullable(resultType) && (returnTypeIsNonNullable || returnTypeIsUnconstrainedTypeParameter) ||
+                        IsUnconstrainedTypeParameter(resultType?.TypeSymbol) && returnTypeIsNonNullable)
```

Comment:
>returnTypeIsNonNullable [](start = 80, length = 23)

What if return type is an unconstrained type parameter? Perhaps a different one.

Reply:
We should have a `Conversion` in that case, with warnings reported in `VisitConversion`.

---
In reply to: [177538120](https://github.com/dotnet/roslyn/pull/25733#discussion_r177538120) [](ancestors = 177538120)